docs/blueprints/EP035.rst

```diff
@@ -0,0 +1,208 @@
+``GenericEntity`` called ``status_reason``,
+and adding in ``status_reason_func``
+to produce values for ``status_reason``.
+The produced ``status_reason`` will be a set of strings,
+so all possible reasons a device might be disable
+can be easily represented.
+
+Example usage:
+
+.. code-block:: python3
+
+    from kytos.core.switch import Switch
+    maintenance_switches = set()
+
+    def switch_maintenance_status(switch):
+        if switch.id in maintenance_switches:
+            return EntityStatus.DOWN
+        return EntityStatus.UP
+
+    Switch.register_status_func('maintenance', switch_maintenance_status)
+
+    def switch_maintenance_reason(switch):
+        if switch.id in maintenance_switches:
+            return {'maintenance'}
+        return set()
+
+    Switch.register_status_reason_func('maintenance', switch_maintenance_reason)
+
+
+New Topology REST Endpoints
+###########################
+
+The following set of new endpoints are intended to allow
+for previewing the affects of adding in interruptions.
+
+
+REST Endpoint ``GET /topology/v3/preview_interruption``
+-------------------------------------------------------
+
+For all know service providers,
+previews affected if the specified interruption occurs now.
+
+Expected Input:
+
+.. code-block:: http
+
+    GET /topology/v3/preview_interruption?switch=<switch_id>&link=<link_id>&interface=<interface_id>&type=<interruption_type>
+
+
+Expected Response:
+
+.. code-block:: json
+
+    {
+        <provider name str>: [<service name str>],
+    }
+
+Side Effects:
+
+- None
+
+Notes:
+
+- Should support multiple key-value pairs with the same key in the query parameters.
+
+
+Endpoint: ``GET <url for previewing>``
+--------------------------------------
+
+For the given service provider,
+previews services affected if the specified interruption occurs now.
+
+Expected Input:
+
+.. code-block:: http
+
+    GET <preview_url>?switch=<switch_id>&link=<link_id>&interface=<interface_id>&type=<interruption_type>
+
+Expected Output:
+
+.. code-block:: json
+
+    [
+        {
+            'id': <service name str>,
+        },
+    ]
+
+Side Effects:
+
+- None
+
+Notes:
+
+- Should support multiple key-value pairs with the same key in the query parameters.
+
+New Kytos Events
+################
+
+The following set of event listeners are intended to be used to notify
+other NApps when a change to the status of a set of devices occurs.
+
+KytosEvent ``topology.interruptions.start``
+-------------------------------------------
+
+Notify services of an interruption starting,
+after the interruption provider has already updated the status.
+
+Expected Input:
+
+.. code-block:: python3
+
+    {
+        "type": <interruption type str>,
+        "switches": [<switch id>],
+        "interfaces": [<interface id>],
+        "links": [<link id>],
+    }
+
+Side Effects:
+
+- Notifies the other NApps of the start of the interruption on the selected devices.
```

Comment:
Exceptionally for MW, we need to confirm if an actual shutdown is needed or not as a side effect so we might need to have a way to parametrize this. Will network engineers actually expect that a interruption of a MW actually also set interfaces physically down too or not? If they do, we need to make it clear that they'll need to undo the shutdown if mid maintenance they also want to run tests.

Reply:
Starting an interruption should neither disable nor deactivate a device, it should only set the status and status reason. This allows NApps to ignore interruptions by checking the status reason. If we add in a new parameter to `mef_eline` which specifies interruption types to ignore, we can then create paths for testing while an interruption is occurring.

Reply:
> Starting an interruption should neither disable nor deactivate a device, it should only set the status and status reason. This allows NApps to ignore interruptions by checking the status reason.

Agreed, this is for sure the way that maximizes composability and minimizes side effects and surprises and easier to maintain. However, the question still remains: will network engineers actually expect that a interruption of a MW actually also set interfaces physically down too or not? 

This is a business requirement that needed to be sorted out with @italovalcy and @jab1982, can you sort that out? If they don't expect for interfaces to be shutdown, then it facilitates, but if they do, then we'll need to find a balance if they'll perform the shutdown via topology API or if we'll need to cover additional side effects.

Reply:
> I chatted with @italovalcy. According to him maintenance is just supposed to make it so devices cannot participate in any services. He says that setting a device down isn't really semantically correct, but would ultimately want to have custom statuses. I told him that status reason basically implements customized statuses with the ability to be ignored, and he liked that.

Excellent. Great that as it is without actual putting devices down is sufficient for the production use case. Let's mark this conversation as resolved then.

Reply:
> @viniarck do you know if any NApps treat EnittyStatus.DOWN and EnittyStatus.DISABLED differently?

On our kytos-ng org, no. We mostly have either conditionals for `status == EntityStatus.UP` or `status != EntityStatus.UP`. However, `status` has been recently exposed on `as_dict()` for every `GenericEntity` instances. So, still, they carry that semantics for any of the disabling or deactivation that historically weren't associated with adding a reason (that has been around for years). 

> Its conceivable that we could replace status entirely with status reasons by making EntityStatus.UP an empty frozen set, making all previous usage of status == EntityStatus.UP still valid, though status is EntityStatus.UP may become invalid.

Good observation. You can create a low priority issue to capture this idea for a future discussion. As of now, I don't think it's worth pursing that route, mainly due to much refactoring and breaking APIs. Let's keep evolving with `status_reason`, when the time comes we can reassess this again, OK?